python/lsst/ctrl/bps/htcondor/handlers.py

```diff
@@ -0,0 +1,252 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Definitions of handlers of HTCondor job ClassAds."""
+
+__all__ = [
+    "Chain",
+    "Handler",
+    "JobCompletedWithToeHandler",
+    "JobCompletedWithoutToeHandler",
+    "JobHeldByOtherHandler",
+    "JobHeldBySignalHandler",
+    "JobHeldByUserHandler",
+]
+
+
+import abc
+import logging
+import re
+from collections.abc import Sequence
+from typing import Any
+
+_LOG = logging.getLogger(__name__)
+
+
+class Handler(abc.ABC):
+    """Abstract base class defining Handler interface."""
+
+    @abc.abstractmethod
+    def handle(self, ad: dict[str, Any]) -> dict[str, Any] | None:
+        """Handle a job ClassAd.
+
+        Parameters
+        ----------
+        ad : `dict[`str`, Any]`
+            The dictionary representing job ClassAd that need to be processed.
+
+        Returns
+        -------
+        ad : `dict[`str`, Any]` | None
+            The dictionary representing job's ClassAd after processing.
+        """
+
+
+class Chain(Sequence):
+    """Class defining chaining of handlers.
+
+    Parameters
+    ----------
+    handlers : `Sequence` [`Handler`]
+        List of handlers that will be used to initialize the chain.
+    """
+
+    def __init__(self, handlers: Sequence[Handler] = None) -> None:
+        self._handlers = []
+        if handlers is not None:
+            for handler in handlers:
+                self.append(handler)
+
+    def __getitem__(self, index: int) -> Handler:
+        return self._handlers[index]
+
+    def __len__(self) -> int:
+        return len(self._handlers)
+
+    def append(self, handler: Handler) -> None:
+        """Append a handler to the chain.
+
+        Parameters
+        ----------
+        handler : `Handler`
+            The handler that needs to be added to the chain.
+
+        Raises
+        ------
+        TypeError
+            Raised if the passed object in not a ``Handler``.
+        """
+        if not isinstance(handler, Handler):
+            raise TypeError(f"unsupported operand type {type(handler)}")
+        self._handlers.append(handler)
+
+    def handle(self, ad: dict[str, Any]) -> dict[str, Any] | None:
+        """Handle a job ClassAd.
+
+        Parameters
+        ----------
+        ad : `dict[`str`, Any]`
+            The dictionary representing a job ClassAd that need to be handled.
+
+        Returns
+        -------
+        ad : `dict[`str`, Any]`
+            A modified job ClassAd if any handler in the chain was able to
+            process the ad, None otherwise.
+        """
+        new_ad = None
+        for handler in self:
+            try:
+                new_ad = handler.handle(ad)
+            except KeyError as e:
+                _LOG.debug(
+                    "Handler %s failed to process the ad for job '%s.%s': required attribute %s missing. "
```

Comment:
If a handler is not able to process an ad, it returns `None`. The try-except is here to catch any failures while a handler is processing an ad so the attempts to process it can continue (as there may be other handlers in the chain able to handle the ad). I fine with changing the logging level (i.e. to `WARNING`) and/or slightly rewording the message to make this more clear, but I don't think the current message is misleading.